compiler/parser/RdrHsSyn.hs

```diff
@@ -2261,9 +2275,29 @@ mkUpdate = mkAppTy (mkQualType "Update")
 partiesType :: LHsType GhcPs
 partiesType = noLoc $ HsListTy noExt $ mkQualType "Party"
 
+mkAppTyArgs :: LHsType GhcPs -> [Located RdrName] -> LHsType GhcPs
+mkAppTyArgs tyCon tyVars = foldl' mkAppTy tyCon $ map rdrNameToType tyVars
+
+mkTemplateType :: String -> [Located RdrName] -> LHsType GhcPs
+mkTemplateType templateName tyVars = mkAppTyArgs (mkUnqualType templateName) tyVars
+
+mkChoiceType :: Located RdrName -> [Located RdrName] -> LHsType GhcPs
+mkChoiceType choiceName tyVars =
+  case occNameString $ rdrNameOcc $ unLoc choiceName of
+    "Archive" -> rdrNameToType choiceName
```

Comment:
Hmm I'm not sure how this would impact the error message. This just decides whether to drop the type parameters in occurrences of the choice data type. It should give the same error of ambiguous data types called `Archive`. I don't mind making the check more specific, but we'll have to this code anyway if we choose your approach of dropping unused type parameters.